# Analyzing the development of Data Science in Europe

In the scope of this Udacity project, I would like to focus on the development of Data Science jobs in European countries. Therefor, an annually Stack Overflow developer survey from 2011 until 2018 provides a wide range of data to tackle this topic and to find interesting insights within the trend of Data Science in Europe.

This Jupyter Notebook contains the code and some descriptions required within the process of analyzing the temporal growth and development of the Data Science community in Europe in the recent years. Do European countries go through the same development like American or Asian countries? I am also interested in the opinion of other Data Scientist according to their job and career satisfaction as well as their salary. What are influential and decisive features relating to these factors?

The structur of this file follows the CRISP-DM (Cross-Industry Standard Process for Data Mining) process:

1. Data Understanding
First, we want to understand the provided data. Therefor, we import the Stack Overview developer survey data and take a look at the provided information within this database.

2. Business Understanding
Second, according to our defined goals from above, we want to formulate appropriate questions. These will help us to tackle the field in a structured manner.

3. Data Preparation
In the third step, we will prepare the data to obtain suitable variables for visualizing the information and answering the formulated questions.

4. Modeling
The fourth step provides the implementation of Machine Learning tools and algorithms to create predictive models for our data. Here, we want to find information about influential elements for our desired variables (satisfaction, salary).

5. Evaluation
In this step, we want to evaluate the developed model.

6. Deployment
The presentation and deployment of the obtained insights will be done within a blog post on Medium (see XXX)

## 1. Data Understanding
The data provided by the Stack Overview developer survey can be downloaded at [https://insights.stackoverflow.com/survey](https://insights.stackoverflow.com/survey). Here, we can find the survey data from 2011 until 2018 downloadable in several zip-files. From 2011 to 2016, these files only contained the data in CSV-files (Comma-Separated Vectors) and, intermittently, a ReadMe-File defining the variables. Since 2017, a PDF-file containing the asked questions and possible responses as well as another CSV-file containing the schema of the data has been additionally provided.

Beforehand, I downloaded these files and saved them into a 'data' folder, including all data CSV-files, and a 'schema' folder with the schema files as well as the pdf files. The files are structured as follows:
- survey_data_201X.csv
- survey_schema_201X.csv

At first, have to import the required packages and the data for the analysis process.

In [266]:
# import packages
import pandas as pd
pd.set_option('display.max_columns', 150)
pd.set_option('display.width', 1000)
pd.options.mode.chained_assignment = None
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

The data is not consistent over all years of the survey. There is a different amount of questions, different variables and, hence, datasets with different shapes. To store everything in one variable, I decided to create a dictionary with the keys indicating the year and values holding Pandas DataFrames.

Since some error messages and warnings occurred, we have to set the encoding and low memory parameters of the read_csv() function of Pandas. Furthermore, the header is individually set for each year and has to be defined in the parameters.

In [267]:
# read Stack Overflow developer survey data from different years and store into a dictionary
data = {}
data['2011'] = pd.read_csv('./data/survey_data_2011.csv', encoding='latin_1', low_memory=False, header = [0,1])
data['2012'] = pd.read_csv('./data/survey_data_2012.csv', encoding='latin_1', low_memory=False, header = [0,1])
data['2013'] = pd.read_csv('./data/survey_data_2013.csv', encoding='latin_1', low_memory=False, header = [0,1])
data['2014'] = pd.read_csv('./data/survey_data_2014.csv', encoding='latin_1', low_memory=False, header = [0,1])
data['2015'] = pd.read_csv('./data/survey_data_2015.csv', encoding='latin_1', low_memory=False, skiprows=1)
data['2016'] = pd.read_csv('./data/survey_data_2016.csv', encoding='latin_1', low_memory=False)
data['2017'] = pd.read_csv('./data/survey_data_2017.csv', encoding='latin_1', low_memory=False)
data['2018'] = pd.read_csv('./data/survey_data_2018.csv', encoding='latin_1', low_memory=False)

In [268]:
# import schema for 2017 and 2018
schema = {}
schema['2017'] = pd.read_csv('./schema/survey_schema_2017.csv')
schema['2018'] = pd.read_csv('./schema/survey_schema_2018.csv')

Let's take a look at the structure and shape of each dataframe.

In [269]:
# print out the amount of rows and columns and the type of columns to explore the datasets
print('year \t rows   \t columns \t type of columns')
print(90*'-')
for i in data.keys():
    print('{} \t {}   \t {} \t\t {}'.format(i,data[i].shape[0],data[i].shape[1],type(data[i].columns)))

year 	 rows   	 columns 	 type of columns
------------------------------------------------------------------------------------------
2011 	 2813   	 65 		 <class 'pandas.core.indexes.multi.MultiIndex'>
2012 	 6243   	 75 		 <class 'pandas.core.indexes.multi.MultiIndex'>
2013 	 9742   	 128 		 <class 'pandas.core.indexes.multi.MultiIndex'>
2014 	 7643   	 120 		 <class 'pandas.core.indexes.multi.MultiIndex'>
2015 	 26086   	 222 		 <class 'pandas.core.indexes.base.Index'>
2016 	 56030   	 66 		 <class 'pandas.core.indexes.base.Index'>
2017 	 51392   	 154 		 <class 'pandas.core.indexes.base.Index'>
2018 	 98855   	 129 		 <class 'pandas.core.indexes.base.Index'>


Obviously, the survey was performed each year using a different amount of questions. The number of participants has also increased from 2,814 in 2011 to nearly 100,000 in 2018.
To compare the survey in 2011 and 2018, we take a look at the columns and a data sample.

In [270]:
# data sample of 2011
data['2011'].head()

What Country or Region do you live in? Which US State or Territory do you live in? How old are you? How many years of IT/Programming experience do you have? How would you best describe the industry you work in? Which best describes the size of your company? Which of the following best describes your occupation? How likely is it that a recommendation you make will be acted upon? What is your involvement in purchasing? You can choose more than 1. Unnamed: 9_level_0 Unnamed: 10_level_0 Unnamed: 11_level_0 Unnamed: 12_level_0 Unnamed: 13_level_0 Unnamed: 14_level_0 What types of purchases are you involved in? Unnamed: 16_level_0 Unnamed: 17_level_0                     Unnamed: 18_level_0 Unnamed: 19_level_0 Unnamed: 20_level_0 What is your budget for outside expenditures (hardware, software, consulting, etc) for 2011? Unnamed: 22_level_0 Unnamed: 23_level_0 Unnamed: 24_level_0 Unnamed: 25_level_0  Unnamed: 26_level_0 Unnamed: 27_level_0 Unnamed: 28_level_0  \
                                Response                                    Response         Response                                                 Response                                              Response                                       Response                                               Response                                                            Response                                                          Influencer        Recommender            Approver           Purchaser        Check Writer      No Involvement        I'm a Seller                                     Hardware             Servers            Software  User Equipment: Monitors, PCs, Laptops         Consultants               Other                                                                                     <$10,000   $10,001 - $25,000   $25,001 - $40,000   $41,000 - $75,000  $75,001 - $100,000  $100,001 - $150,000           >$150,000          Don't know   
0                                 Africa                                         NaN             < 20                                                 <2                                               Consulting                                   Start Up (1-25)                          Web Application Developer                                 Not in a million years                                                                 NaN                                 NaN            Approver                 NaN                 NaN                 NaN                 NaN                                          NaN                 NaN            Software                                     NaN                 NaN                 NaN                                                NaN                                                           NaN   $25,001 - $40,000                 NaN                 NaN                  NaN                 NaN                 NaN   
1                           Other Europe                                         NaN            25-29                                              41310                                        Software Products                    Mature Small Business (25-100)                                  Server Programmer                              It's been known to happen                                                                 NaN                                 NaN                 NaN                 NaN                 NaN      No Involvement                 NaN                                          NaN                 NaN                 NaN                                     NaN                 NaN                 NaN                                                NaN                                                           NaN                 NaN                 NaN                 NaN                  NaN                 NaN                 NaN   
2                                  India                                         NaN            25-29             

In [271]:
# data sample of 2018
data['2018'].head()

Respondent Hobby OpenSource         Country         Student          Employment                                    FormalEducation                                     UndergradMajor               CompanySize                                            DevType       YearsCoding YearsCodingProf                     JobSatisfaction                  CareerSatisfaction                                      HopeFiveYears                                    JobSearchStatus                 LastNewJob  AssessJob1  AssessJob2  AssessJob3  AssessJob4  AssessJob5  AssessJob6  AssessJob7  AssessJob8  AssessJob9  AssessJob10  AssessBenefits1  AssessBenefits2  AssessBenefits3  AssessBenefits4  AssessBenefits5  AssessBenefits6  AssessBenefits7  AssessBenefits8  AssessBenefits9  AssessBenefits10  AssessBenefits11  JobContactPriorities1  JobContactPriorities2  JobContactPriorities3  JobContactPriorities4  JobContactPriorities5  JobEmailPriorities1  JobEmailPriorities2  JobEmailPriorities3  \
0           1   Yes         No           Kenya              No  Employed part-time         Bachelor’s degree (BA, BS, B.Eng., etc.)                          Mathematics or statistics        20 to 99 employees                               Full-stack developer         3-5 years       3-5 years                 Extremely satisfied                 Extremely satisfied  Working as a founder or co-founder of my own c...  I’m not actively looking, but I am open to n...       Less than a year ago        10.0         7.0         8.0         1.0         2.0         5.0         3.0         4.0         9.0          6.0              NaN              NaN              NaN              NaN              NaN              NaN              NaN              NaN              NaN               NaN               NaN                    3.0                    1.0                    4.0                    2.0                    5.0                  5.0                  6.0                  7.0   
1           3   Yes        Yes  United Kingdom              No  Employed full-time         Bachelor’s degree (BA, BS, B.Eng., etc.)  A natural science (ex. biology, chemistry, phy...  10,000 or more employees  Database administrator;DevOps specialist;Full-...  30 or more years     18-20 years             Moderately dissatisfied  Neither satisfied nor dissatisfied  Working in a different or more specialized tec...                    I am actively looking for a job      More than 4 years ago         1.0         7.0        10.0         8.0         2.0         5.0         4.0         3.0         6.0          9.0              1.0              5.0              3.0              7.0             10.0              4.0             11.0              9.0              6.0               2.0               8.0                    3.0                    1.0                    5.0                    2.0                    4.0                  1.0                  3.0                  4.0   
2           4   Yes        Yes   United States              No  Employed full-time                                   Associate degree  Computer science, computer engineering, or sof...        20 to 99 employees           Engineering manager;Full-stack developer       24-26 years       6-8 years                Moderately satisfied                Moderately satisfied  Working as a founder or co-founder of my own c...  I’m not actively looking, but I am open to n...       Less than a year ago         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN          NaN              NaN              NaN              NaN              NaN              NaN              NaN              NaN              NaN              NaN               NaN               NaN                    NaN                    NaN                    NaN                    NaN                    NaN                  NaN                  NaN                  NaN   
3           5    No         No   United States     

In [272]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(schema['2018'])

                          Column                                       QuestionText
0                     Respondent  Randomized respondent ID number (not in order ...
1                          Hobby                            Do you code as a hobby?
2                     OpenSource         Do you contribute to open source projects?
3                        Country          In which country do you currently reside?
4                        Student  Are you currently enrolled in a formal, degree...
5                     Employment  Which of the following best describes your cur...
6                FormalEducation  Which of the following best describes the high...
7                 UndergradMajor  You previously indicated that you went to a co...
8                    CompanySize  Approximately how many people are employed by ...
9                        DevType  Which of the following describe you? Please se...
10                   YearsCoding  Including any education, for how many year

In [273]:
# to get a better data understanding, a complete description of the variables is needed
# Therefor, we define the get_description()-function from the Udacity "How To Break Into the Field"-lesson
def get_description(column_name, schema_year = '2018', schema = schema):
    '''
    INPUT - schema - pandas dataframe with the schema of the developers survey
            column_name - string - the name of the column you would like to know about
    OUTPUT - 
            desc - string - the description of the column
    '''
    schema = schema[schema_year]
    if schema_year == '2018':
        desc = list(schema[schema['Column'] == column_name]['QuestionText'])[0]
    elif schema_year == '2017':
        desc = list(schema[schema['Column'] == column_name]['Question'])[0]
    
    return desc

get_description('CompanyType', '2017')

'Which of the following best describes the type of company or organization you work for?'

## 2. Business Understanding
In this project, I am interested in the trend of Data Science jobs in European countries. With the support of the data provided by the Stack Overflow survey, I would like to answer these questions:

1. Analyzing Trends and Developments in Data Science jobs
    - What is the chronological trend of Data Science jobs in Europe in comparison to American, Asian or African countries?
    - How are data scientists actually distributed over continents and countries?


2. What are common job requirements and qualifications for data science jobs in Europe?
    - What is the common level of education of employees in the field of data science?
    - How old are these data scientists?
    - Are there more male or female data scientists in Europe?
    - How many years of coding experience do data scientists in Europe have?
    - In what company sizes do data scientists typically participate in Europe?
    - How are data scientists employed within companies?
    - What tools are commonly used by data scientists in Europe?


3. Which features mostly influence the job satisfaction according to data scientists in Europe?


4. Which features mostly influence the salary of data scientists in Europe?

## 3. Data Preparation
According to the table in section one (Data Understanding) and the shown data samples, the survey datasets have different amount of columns, different types of column headers and names as well as different variables. E.g., we cannot compare the annual surveys on every variable without data preparation. Furthermore, the datasets contain lots of missing values, that will result in errors within the modeling process. Thus, this section will be structured into several individual problems, that have to be handled to obtain a usable dataset.

### 3.1 Select required variables and gather data
In the first data preparation step, we want to select the information, we need to answer the questions stated in section two (Business Understanding). Here, we define the required information for each set of question and name the corresponding variable from the datasets in brackets.

- Required for question 1:
    - to obtain a chronological development, data from 2011 to 2018 are required
    - Occupation (named differentially in the datasets)
    - Region (named differentially in the datasets or has to be created)
    - Country (named differentially in the datasets)
    

- Required for question 2:
    - to obtain an actual impression, we only need survey data from 2018 and from European countries
    - Level of Education (FormalEducation)
    - Age (Age)
    - Gender (Gender)
    - Programming Experience (YearsCoding)
    - Programming language required (LanguageWorkedWith)
    - Size of Company (CompanySize)
    - Status of Employment (Employment and Student)
    - Main Field of Studies (UndergradMajor)
    - Occupation (DevType)
    - Country (Country)
    - Non-degree Education (EducationTypes)
    - Satisfaction (JobSatisfaction, CareerSatisfaction)
    - Salary (Salary)


- Required for question 3 and 4:
    - to obtain an actual impression, we only need survey data from 2018 and from European countries
    - input features:
        - variables of question 2
    
    - response variables:
        - Salary (Salary)
        - Satisfaction (JobSatisfaction, CareerSatisfaction)
    

For question 3 and 4, we want to model the salary as well as job and career satisfaction of data scientists in European countries depending on the input features, we explored in question. Therefor, we ignore variables regarding for example the usage of Stack Overflow. We would like to focus on features concerning the education, qualification and employment status as well as main characteristics of each data scientist (gender, age etc.).

#### 3.1.1 Adjust geographical information

We want to focus inter alia on the geographical distribution of data science and its change over time. Furthermore, we want to focus on European data scientists, thus, we have to filter the data on the country or region of these survey participants. Therefor, we require consitent regional data.

In this step, we unify the country names and assign their specific and superordinate region using a dictionary to find geographical characteristics.

In [274]:
# create a region dictionary
region_dict = {}
for country in data['2016'].country.unique():
    regions = data['2016'][data['2016'].country==country].so_region.unique()
    if regions.size>0:
        region_dict[country] = regions[0]

region_dict['Congo {Democratic Rep}'] = 'Africa'
region_dict['Congo, Republic of the...'] = 'Africa'
region_dict['Libya'] = 'Africa'
region_dict['Eritrea'] = 'Africa'
region_dict['Mali'] = 'Africa'
region_dict['Swaziland'] = 'Africa'
region_dict['Gambia'] = 'Africa'
region_dict['Senegal'] = 'Africa'
region_dict['Cape Verde'] = 'Africa'
region_dict['South Sudan'] = 'Africa'
region_dict['Burkina Faso'] = 'Africa'
region_dict['Reunion (French)'] = 'Africa'
region_dict['Comoros'] = 'Africa'
region_dict['Central African Republic'] = 'Africa'
region_dict['Liberia'] = 'Africa'
region_dict['Guinea-Bissau'] = 'Africa'
region_dict['Lesotho'] = 'Africa'
region_dict['South America'] = 'South America'
region_dict['S. Georgia & S. Sandwich Isls.'] = 'South America'
region_dict['French Guyana'] = 'South America'
region_dict['Falkland Islands'] = 'South America'
region_dict['Azerbaidjan'] = 'Eastern Europe'
region_dict['Central America'] = 'Central America'
region_dict['North America'] = 'North America'
region_dict['Africa'] = 'Africa'
region_dict['Solomon Islands'] = 'Oceania'
region_dict['American Samoa'] = 'Oceania'
region_dict['Virgin Islands (USA)'] = 'Oceania'
region_dict['Virgin Islands (British)'] = 'Oceania'
region_dict['Middle East'] = 'Middle East'
region_dict['Brunei'] = 'Southeast Asia'
region_dict['Christmas Island'] = 'Southeast Asia'
region_dict['Timor-Leste'] = 'Southeast Asia'
region_dict['Turkmenistan'] = 'Central Asia'
region_dict['Tadjikistan'] = 'Central Asia'
region_dict['Netherlands Antilles'] = 'Caribbean'
region_dict['Anguilla'] = 'Caribbean'
region_dict['Saint Lucia'] = 'Caribbean'
region_dict['Aruba'] = 'Caribbean'
region_dict['Antigua and Barbuda'] = 'Caribbean'
region_dict['Montserrat'] = 'Caribbean'
region_dict['Martinique (French)'] = 'Caribbean'
region_dict['Saint Vincent & Grenadines'] = 'Caribbean'
region_dict['Cayman Islands'] = 'Caribbean'
region_dict['U.S. Minor Outlying Islands'] = 'Oceania'
region_dict['New Caledonia (French)'] = 'Oceania'
region_dict['Gibraltar'] = 'Western Europe'
region_dict['Micronesia'] = 'Oceania'
region_dict['Polynesia (French)'] = 'Oceania'
region_dict['Pitcairn Island'] = 'Oceania'
region_dict['Guam'] = 'Oceania'
region_dict['Northern Mariana Islands'] = 'Oceania'
region_dict['Cook Islands'] = 'Oceania'
region_dict['Marshall Islands'] = 'Oceania'
region_dict['Nauru'] = 'Oceania'
region_dict['Australasia'] = 'Oceania'

In [275]:
def clean_countries(df, regions=region_dict):
    '''
    unify the country variable and create a new columns with corresponding regions
    
    INPUT - df - Pandas Dataframe with column countries which has to be adjusted
            regions - dictionary containing country - region assignments
            
    OUTPUT - df - adjusted Dataframe with additional 'region' and 'superordinate_region' columns
    '''
    
    # clean double country names
    US_states = ['California', 'New Jersey', 'Nevada', 'Wisconsin','Pennsylvania', 'New York', 'South Carolina', 'Michigan',
                 'Texas', 'Washington', 'Minnesota', 'North Carolina', 'Ohio', 'Illinois', 'Florida', 'Arizona', 'Tennessee',
                 'Colorado', 'Alabama', 'Oregon', 'Kentucky', 'Massachusetts', 'Alaska', 'Idaho', 'South Dakota', 'Missouri',
                 'Maryland', 'New Hampshire', 'Virginia', 'Maine', 'Utah', 'Washington DC', 'Louisiana', 'Indiana', 'Nebraska',
                 'Vermont', 'Iowa', 'Oklahoma', 'Montana', 'Connecticut', 'Kansas', 'Wyoming', 'Arkansas', 'West Virginia',
                 'Delaware', 'Rhode Island', 'New Mexico', 'Mississippi', 'North Dakota', 'United States of America']
    
    Congo = ['Congo {Democratic Rep}', 'Democratic Republic of the Congo', 'Zaire']
    S_Korea = ['South Korea', 'Korea South', 'Republic of Korea']
    N_Korea = ['North Korea', 'Korea North', "Democratic People's Republic of Korea"]
    China = ['China', 'Hong Kong', 'Hong Kong (S.A.R.)', 'Macau']
    
    df['country'] = df['country'].apply(lambda x: 'United States' if x in US_states else x)
    df['country'] = df['country'].apply(lambda x: 'Korea South' if x in S_Korea else x)
    df['country'] = df['country'].apply(lambda x: 'Korea North' if x in N_Korea else x)
    df['country'] = df['country'].apply(lambda x: 'China' if x in China else x)
    df['country'] = df['country'].apply(lambda x: 'Macedonia' if 'macedonia' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Myanmar, {Burma}' if 'myanmar' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Congo {Democratic Rep}' if x in Congo else x)
    df['country'] = df['country'].apply(lambda x: 'Vatican City' if 'vatican' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Bosnia Herzegovina' if 'bosnia' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Venezuela' if 'venezuela' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Iran' if 'iran' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Slovakia' if 'slovak' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Tanzania' if 'tanzania' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Moldova' if 'mold' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Brunei' if 'brunei' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Ivory Coast' if 'ivo' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Vietnam' if 'viet' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Syria' if 'syr' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Ireland {Republic}' if 'ireland' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Russian Federation' if 'russia' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Trinidad & Tobago' if 'trinidad' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Finland' if 'Aland' in str(x) else x)
    df['country'] = df['country'].apply(lambda x: 'Libya' if 'libya' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: 'Micronesia' if 'micronesia' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: np.nan if 'other' in str(x).lower() else x)
    df['country'] = df['country'].apply(lambda x: np.nan if 'i prefer not to say' in str(x).lower() else x)
    
    # create region column based on dictionary
    df['region'] = df['country'].apply(lambda x: regions[x] if x in list(regions.keys()) else np.nan)
    
    Europe = ['Western Europe', 'Eastern Europe']
    America = ['North America', 'Central America', 'South America', 'Caribbean']
    Asia = ['South Asia', 'Middle East', 'Southeast Asia', 'East Asia', 'Central Asia']

    # create subordinate_region columns based on region
    df['superordinate_region'] = df['region'].apply(lambda x: 'Europe' if x in Europe else
                                                           ('America' if x in America else
                                                           ('Asia' if x in Asia else x)))
    # drop all nan values
    df.dropna(axis=0,subset=['country','region','superordinate_region'], inplace = True)
    
    return df

#### 3.1.2 Identify Data Scientists

In various columns, information about the actual job are contained. We would like to find out, how many of the survey participants are working within the field of data science.

We will add another column showing whether the person is a data scientist (1) or not (0).

In [276]:
def add_data_science_job(df):
    '''
    find data science jobs within the occupation column of the dataframe
    
    INPUT - df - Pandas Dataframe with column occupation
            
    OUTPUT - df - Pandas Dataframe with new 'boolean' Data Science column 
                  (1 = job linked with Data Science, 0 = no link to Data Science)
    '''
    
    # create new data_science column with 1 and 0 mapping
    # inspired by: https://github.com/gauravansal/Analysis-of-Data-Science-community-growth-from-2011-2018
    df['data_science'] = df['occupation'].str.contains('.*ata|scientist|machine|learning|algorithm|business|analyst')
    df['data_science'] = df['data_science'].apply(lambda x: 1 if x == True else (0 if x == False else x))
    
    return df

#### 3.1.3 Gathering and Selecting Data for Question Sets 1 and 2

After defining functions for cleaning and adding geographical and regional information as well as identifying data scientists within the group of survey participants, we apply them to define the datasets required for tackling the questions.

We distinguish between chronological data, which filters from all datasets from 2011 to 2018 and the actual data from 2018. Within the latter we will focus on more different features taken from the 2018 Stack Overflow survey.

In [277]:
# create a dataset with required information for question 1
def create_data_question1(data=data):
    '''
    Create a DataFrame containing the occupation and regional data (country) for each year (2011-2018)
   
    INPUT - data - dictionary containg the survey data for each year
    
    OUTPUT - q1_df - Pandas Dataframe containg required columns
    '''
    q1_dict = {}
    
    col_names=['country', 'occupation']
    
    # import data from year 2011
    select_cols_ind = [0, 6]
    q1_dict['2011'] = data['2011'].iloc[:,select_cols_ind]
    q1_dict['2011'].columns = col_names
    q1_dict['2011'].loc[:,'year'] = 2011
    
    # import data from year 2012
    select_cols_ind = [0, 6]
    q1_dict['2012'] = data['2012'].iloc[:,select_cols_ind]
    q1_dict['2012'].columns = col_names
    q1_dict['2012'].loc[:,'year'] = 2012
    
    # import data from year 2013
    select_cols_ind = [0, 6]
    q1_dict['2013'] = data['2013'].iloc[:,select_cols_ind]
    q1_dict['2013'].columns = col_names
    q1_dict['2013'].loc[:,'year'] = 2013
    
    # import data from year 2014
    select_cols_ind = [0, 6]
    q1_dict['2014'] = data['2014'].iloc[:,select_cols_ind]
    q1_dict['2014'].columns = col_names
    q1_dict['2014'].loc[:,'year'] = 2014
    
    # import data from year 2015
    select_cols_ind = [0, 5]
    q1_dict['2015'] = data['2015'].iloc[:,select_cols_ind]
    q1_dict['2015'].columns = col_names
    q1_dict['2015'].loc[:,'year'] = 2015
    
    # import data from year 2016
    select_cols_ind = [2, 10]
    q1_dict['2016'] = data['2014'].iloc[:,select_cols_ind]
    q1_dict['2016'].columns = col_names
    q1_dict['2016'].loc[:,'year'] = 2016
    
    # import data from year 2017
    # merge the developer type and non-developer type into a new occupation column
    select_cols_ind = [3, 154]
    data['2017'].loc[:,'occuption'] = data['2017'].iloc[:,14:18].apply(lambda x:
                                                                       ";".join([i for i in list(x) if not pd.isna(i)]),axis=1)
    q1_dict['2017'] = data['2017'].iloc[:,select_cols_ind]
    q1_dict['2017'].columns = col_names
    q1_dict['2017'].loc[:,'year'] = 2017
    
    # import data from year 2018
    select_cols_ind = [3, 9]
    q1_dict['2018'] = data['2018'].iloc[:,select_cols_ind]
    q1_dict['2018'].columns = col_names
    q1_dict['2018'].loc[:,'year'] = 2018
    
    # combine DataFrames from dictionary
    q1_df = q1_dict['2011']
    for year in range(2012,2019):
        q1_df = q1_df.append(q1_dict[str(year)],ignore_index=True, sort=False)
        
    # clean and adjust countries
    q1_df = clean_countries(q1_df)
    
    # add data science column
    q1_df = add_data_science_job(q1_df)
    
    # rearrange columns
    q1_df = q1_df[['year','country','region','occupation','data_science']]
    
    return q1_df        

In [278]:
q1_df = create_data_question1(data)
print('Shape of dataset for question 1: {}'.format(q1_df.shape))
print('Number of missing values: {}'.format(q1_df.isna().sum().sum()))
print('Data Science jobs: {} \nOther jobs: {}'.format(q1_df.data_science.value_counts()[1],q1_df.data_science.value_counts()[0]))

Shape of dataset for question 1: (199423, 5)
Number of missing values: 26204
Data Science jobs: 31631 
Other jobs: 154690


In [279]:
q1_df.head()

year        country          region                             occupation  data_science
0  2011         Africa          Africa              Web Application Developer           0.0
2  2011          India      South Asia                      Server Programmer           0.0
3  2011        Germany  Western Europe                                Student           0.0
5  2011      Australia         Oceania              Web Application Developer           0.0
6  2011  United States   North America  Executive (VP of Eng, CTO, CIO, etc.)           0.0

In [280]:
# create a dataset with required information for question 1
def create_data_question2(data=data):
    '''
    Create dataframe required for question 2 with:
        FormalEducation, Age, Gender, YearsCoding, YearsCodingProf, LanguageWorkedWith, DatabaseWorkedWith, PlatformWorkedWith,
        CompanySize, Employment, Student, UndergradMajor, DevType, Country, EducationTypes, JobSatisfaction, 
        CareerSatisfaction, ConvertedSalary
    Filter for European countries
    Create new columns region and data_science
   
    INPUT - data - dictionary containg the survey data for each year
    
    OUTPUT - df - Pandas Dataframe containg required columns
    '''
    # import data from year 2018
    data = data['2018']
    
    col_names=['country', 'occupation','formal_education', 'age', 'gender', 'years_coding', 'language',
               'company_size','employment','student','undergrad_major', 'education_type', 
               'job_satisfaction', 'career_satisfaction', 'salary']
    
    select_cols_ind = [3, 9, 6, 124, 120, 10, 65, 8, 5, 4, 7, 58, 12, 13, 54]
    df = data.iloc[:,select_cols_ind]
    df.columns = col_names
        
    # clean and adjust countries
    df = clean_countries(df)
    
    # add data science column
    df = add_data_science_job(df)
    
    return df        

In [281]:
q2_df = create_data_question2(data)
print('Shape of dataset for question 2: {}'.format(q2_df.shape))
print('Data Science jobs: {} \nOther jobs: {}'.format(q2_df.data_science.value_counts()[1],q2_df.data_science.value_counts()[0]))
print('Number of missing values: {}'.format(q2_df.isna().sum().sum()))

Shape of dataset for question 2: (98357, 18)
Data Science jobs: 21866 
Other jobs: 70163
Number of missing values: 293913


In [282]:
q2_df.head()

country                                         occupation                                   formal_education                age gender      years_coding                                          language              company_size          employment         student                                    undergrad_major                                     education_type                    job_satisfaction                 career_satisfaction   salary          region superordinate_region  data_science
0           Kenya                               Full-stack developer         Bachelor’s degree (BA, BS, B.Eng., etc.)  25 - 34 years old   Male         3-5 years                        JavaScript;Python;HTML;CSS        20 to 99 employees  Employed part-time              No                          Mathematics or statistics  Taught yourself a new language, framework, or ...                 Extremely satisfied                 Extremely satisfied      NaN          Africa               Africa           0.0
1  United Kingdom  Database administrator;DevOps specialist;Full-...         Bachelor’s degree (BA, BS, B.Eng., etc.)  35 - 44 years old   Male  30 or more years                      JavaScript;Python;Bash/Shell  10,000 or more employees  Employed full-time              No  A natural science (ex. biology, chemistry, phy...  Taught yourself a new language, framework, or ...             Moderately dissatisfied  Neither satisfied nor dissatisfied  70841.0  Western Europe               Europe           1.0
2   United States           Engineering manager;Full-stack developer                                   Associate degree                NaN    NaN       24-26 years                                               NaN        20 to 99 employees  Employed full-time              No  Computer science, computer engineering, or sof...                                                NaN                Moderately satisfied                Moderately satisfied      NaN   North America              America           0.0
3   United States                               Full-stack developer         Bachelor’s degree (BA, BS, B.Eng., etc.)  35 - 44 years old   Male       18-20 years  C#;JavaScript;SQL;TypeScript;HTML;CSS;Bash/Shell      100 to 499 employees  Employed full-time              No  Computer science, computer engineering, or sof...  Completed an industry certification program (e...  Neither satisfied nor dissatisfied               Slightly dissatisfied      NaN   North America              America           0.0
4    South Africa  Data or business analyst;Desktop or enterprise...  Some college/university study without earning ...  18 - 24 years old   Male         6-8 years                C;C++;Java;Matlab;R;SQL;Bash/Shell  10,000 or more employees  Employed full-time  Yes, part-time  Computer science, computer engineering, or sof...  Taken a part-time in-person course in programm...                  Slightly satisfied                Moderately satisfied  21426.0          Africa               Africa           1.0

### 3.2 Handling missing data

Partially, we already deleted some missing values within the geographical columns.
In this section, we want to identify other missing values and handle them by removing or imputing the data.

#### 3.2.1 Identifying missing values

In [283]:
# check for missing data in dataset for question 1
for col in q1_df. columns:
    print('{:>12}:{:>12} missing values = {:.2f}%'.format(col, q1_df[col].isnull().sum(),q1_df[col].isnull().sum()/q1_df.shape[0]*100))

        year:           0 missing values = 0.00%
     country:           0 missing values = 0.00%
      region:           0 missing values = 0.00%
  occupation:       13102 missing values = 6.57%
data_science:       13102 missing values = 6.57%


Obviously, the columns data_science and occupation have the same number of missing values, because we derived the data_science column from the other one.
As mentioned, we already tackled missing values in the country and region data. The year column is also complete.

The rate of missing values (6.57%) is ok to remove all missing values. Thus, we can keep over 93% of the observations.

In [284]:
# check for missing data in dataset for question 2
for col in q2_df. columns:
    print('{:>20}:{:>12} missing values = {:.2f}%'.format(col, q2_df[col].isnull().sum(),q2_df[col].isnull().sum()/q2_df.shape[0]*100))

             country:           0 missing values = 0.00%
          occupation:        6328 missing values = 6.43%
    formal_education:        3727 missing values = 3.79%
                 age:       33840 missing values = 34.41%
              gender:       33941 missing values = 34.51%
        years_coding:        4594 missing values = 4.67%
            language:       20086 missing values = 20.42%
        company_size:       26871 missing values = 27.32%
          employment:        3106 missing values = 3.16%
             student:        3529 missing values = 3.59%
     undergrad_major:       19378 missing values = 19.70%
      education_type:       30448 missing values = 30.96%
    job_satisfaction:       29131 missing values = 29.62%
 career_satisfaction:       21909 missing values = 22.27%
              salary:       50697 missing values = 51.54%
              region:           0 missing values = 0.00%
superordinate_region:           0 missing values = 0.00%
        data_science: 

In [285]:
# check for rows with only NaN values
q2_df.isnull().all(axis=1).value_counts()

False    98357
dtype: int64

In the second dataset, we have a lot of more missing values. Especially, the salary column has over 50% of missing data, which is a problem, because we want to use this variable for modeling in section 4. Additionally, age and gender have over one third of missing values.

We have no rows that only consist of NaN values.

#### 3.2.2 Tackling missing values

Dataset q1_df only showed some missing data, thus it is acceptable to remove these rows.

In [289]:
q1_df.dropna(axis=0, inplace=True)
print("Missing values in dataset: {}".format(q1_df.isnull().sum().sum()))

Missing values in dataset: 0


Dataset q2_df is a bit more complicated.
At first, we will focus on the variable with the most missing data - "Salary". Here, we will impute the missing values by applying the median of the corresponding country. Because the mean can be shifted to higher values due to very high salaries, we use the median to get a better metric for the salary of data scientists or other developers over the world.

In [292]:
# impute missing salary data by filling in the median of the corresponding country
q2_df["salary"].fillna(q2_df.groupby("country")["salary"].transform("median"), inplace=True) 
print("Missing values in column 'salary' of dataset 2: {}".format(q2_df['salary'].isnull().sum()))

Missing values in column 'salary' of dataset 2: 73


We managed to reduce the missing values in the salary column from over 50,000 to only 73.

In [126]:
q2_df[q2_df.country=='Germany'].age.value_counts().sort_values(ascending=False).index[0]

'25 - 34 years old'

In [203]:
mc = q2_df.groupby(['country'])['age'].value_counts()

In [122]:
# remove all NaNs and show shape
q1_df.dropna(axis=0, inplace=True)
print('Shape of dataset for question 1 after removing NaNs: {}'.format(q1_df.isna().shape))

Shape of dataset for question 1 after removing NaNs: (186321, 5)


In [204]:
mc

country              age               
Afghanistan          18 - 24 years old       6
                     25 - 34 years old       6
                     35 - 44 years old       2
                     45 - 54 years old       1
                     Under 18 years old      1
Albania              25 - 34 years old      21
                     18 - 24 years old      20
                     Under 18 years old      4
                     35 - 44 years old       3
Algeria              25 - 34 years old      27
                     18 - 24 years old      22
                     35 - 44 years old       6
                     45 - 54 years old       3
                     Under 18 years old      2
Andorra              35 - 44 years old       2
                     45 - 54 years old       1
                     55 - 64 years old       1
                     65 years or older       1
Angola               25 - 34 years old       1
Antigua and Barbuda  55 - 64 years old       1
Argentina           

In [133]:
q2_df.head(20)

country                                         occupation                                   formal_education                age  gender      years_coding years_coding_prof                                           language                                           database                                     platform              company_size          employment         student                                    undergrad_major                                     education_type                    job_satisfaction                 career_satisfaction    salary          region superordinate_region  data_science
0            Kenya                               Full-stack developer         Bachelor’s degree (BA, BS, B.Eng., etc.)  25 - 34 years old    Male         3-5 years         3-5 years                         JavaScript;Python;HTML;CSS  Redis;SQL Server;MySQL;PostgreSQL;Amazon RDS/A...                     AWS;Azure;Linux;Firebase        20 to 99 employees  Employed part-time              No                          Mathematics or statistics  Taught yourself a new language, framework, or ...                 Extremely satisfied                 Extremely satisfied   12000.0          Africa               Africa           0.0
1   United Kingdom  Database administrator;DevOps specialist;Full-...         Bachelor’s degree (BA, BS, B.Eng., etc.)  35 - 44 years old    Male  30 or more years       18-20 years                       JavaScript;Python;Bash/Shell                         Redis;PostgreSQL;Memcached                                        Linux  10,000 or more employees  Employed full-time              No  A natural science (ex. biology, chemistry, phy...  Taught yourself a new language, framework, or ...             Moderately dissatisfied  Neither satisfied nor dissatisfied   70841.0  Western Europe               Europe           1.0
2    United States           Engineering manager;Full-stack developer                                   Associate degree                NaN     NaN       24-26 years         6-8 years                                                NaN                                                NaN                                          NaN        20 to 99 employees  Employed full-time              No  Computer science, computer engineering, or sof...                                                NaN                Moderately satisfied                Moderately satisfied  100000.0   North America              America           0.0
3    United States                               Full-stack developer         Bachelor’s degree (BA, BS, B.Eng., etc.)  35 - 44 years old    Male       18-20 years       12-14 years   C#;JavaScript;SQL;TypeScript;HTML;CSS;Bash/Shell  SQL Server;Microsoft Azure (Tables, CosmosDB, ...                                        Azure      100 to 499 employees  Employed full-time              No  Computer science, computer engineering, or sof...  Completed an industry certification program (e...  Neither satisfied nor dissatisfied               Slightly dissatisfied  100000.0   North America              America           0.0
4     South Africa  Data or business analyst;Desktop or enterprise...  Some college/university study without earning ...  18 - 24 years old    Male         6-8 years         0-2 years                 C;C++;Java;Matlab;R;SQL;Bash/Shell               SQL Server;PostgreSQL;Oracle;IBM Db2            Arduino;Windows Desktop or Server  10,000 or more employees  Employed full-time  Yes, part-time  Computer science, computer engineering, or sof...  Taken a part-time in-person course in programm...                  Slightly satisfied                Moderately satisfied   21426.0          Africa               Africa           1.0
5   United Kingdom  Back-end developer;Database administrator;Fron...         Bachelor’s degree (BA, BS, B.Eng., etc.)  18 - 24 years old    Male         6-8 years         3-5 years         Java;JavaScript;Python;TypeScript;HTML;CSS         

C:\Users\Kroening\Anaconda3\lib\site-packages\pandas\core\generic.py:5430: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


In [105]:
q2_df.head()

country                                         occupation                                   formal_education                age gender      years_coding years_coding_prof                                          language                                           database                           platform              company_size          employment         student                                    undergrad_major                                     education_type                    job_satisfaction                 career_satisfaction         salary          region superordinate_region  data_science
0           Kenya                               Full-stack developer         Bachelor’s degree (BA, BS, B.Eng., etc.)  25 - 34 years old   Male         3-5 years         3-5 years                        JavaScript;Python;HTML;CSS  Redis;SQL Server;MySQL;PostgreSQL;Amazon RDS/A...           AWS;Azure;Linux;Firebase        20 to 99 employees  Employed part-time              No                          Mathematics or statistics  Taught yourself a new language, framework, or ...                 Extremely satisfied                 Extremely satisfied   30356.666667          Africa               Africa           0.0
1  United Kingdom  Database administrator;DevOps specialist;Full-...         Bachelor’s degree (BA, BS, B.Eng., etc.)  35 - 44 years old   Male  30 or more years       18-20 years                      JavaScript;Python;Bash/Shell                         Redis;PostgreSQL;Memcached                              Linux  10,000 or more employees  Employed full-time              No  A natural science (ex. biology, chemistry, phy...  Taught yourself a new language, framework, or ...             Moderately dissatisfied  Neither satisfied nor dissatisfied   70841.000000  Western Europe               Europe           1.0
2   United States           Engineering manager;Full-stack developer                                   Associate degree                NaN    NaN       24-26 years         6-8 years                                               NaN                                                NaN                                NaN        20 to 99 employees  Employed full-time              No  Computer science, computer engineering, or sof...                                                NaN                Moderately satisfied                Moderately satisfied  154413.128506   North America              America           0.0
3   United States                               Full-stack developer         Bachelor’s degree (BA, BS, B.Eng., etc.)  35 - 44 years old   Male       18-20 years       12-14 years  C#;JavaScript;SQL;TypeScript;HTML;CSS;Bash/Shell  SQL Server;Microsoft Azure (Tables, CosmosDB, ...                              Azure      100 to 499 employees  Employed full-time              No  Computer science, computer engineering, or sof...  Completed an industry certification program (e...  Neither satisfied nor dissatisfied               Slightly dissatisfied  154413.128506   North America              America           0.0
4    South Africa  Data or business analyst;Desktop or enterprise...  Some college/university study without earning ...  18 - 24 years old   Male         6-8 years         0-2 years                C;C++;Java;Matlab;R;SQL;Bash/Shell               SQL Server;PostgreSQL;Oracle;IBM Db2  Arduino;Windows Desktop or Server  10,000 or more employees  Employed full-time  Yes, part-time  Computer science, computer engineering, or sof...  Taken a part-time in-person course in programm...                  Slightly satisfied                Moderately satisfied   21426.000000          Africa               Africa           1.0